couchbaselabs/observability
# Rework "Add Cluster" to support CBS 7.x

Adding a CBS 7.x target to Prometheus is slightly more complicated than a 6.x target because 7.x's /metrics endpoint requires basic authentication, which can only be passed in through prometheus.yml. This means that our existing "have a single file_sd_config directory" approach is only viable if all the clusters have the same credentials. This also means that each cluster must have its own scrape job.

This commit uses a different method, namely rewriting prometheus.yml at runtime to add in a new scrape_config with the requisite configuration. Currently it uses a static_config for the nodes, and can only add clusters and not edit existing ones. The user's existing configuration is preserved in place as much as possible (including comments), we will only touch scrape_configs that we ourselves have written.

This simplistic implementation is a stepping stone to the full config service which would use http_sd to dynamically update the targets list - it'd still need to add and remove targets to cope with authentication.

Test Plan:
* Start a CBS 7 cluster
* `make container` off this branch
* Run it, port-forwarding 8080
* Go to localhost:8080, click "Add Cluster", and fill in the details
* Check the Prometheus targets page and see that a new job has appeared
* Go to Grafana, open the cluster inventory, and see that the cluster has appeared (this means that cbmultimanager has picked it up)

Note: the single-cluster overview won't work as that assumes 6.x metrics. Another PR is in progress to add a 7.x equivalent.
